client.go

```diff
@@ -8,137 +8,162 @@ import (
+	apiVersion   Version
+	useCDN       bool
+	baseAPIURL   url.URL
+	baseQueryURL url.URL // if useCDN=false, baseQueryURL will be same as baseAPIURL.
+	token        string
+	projectID    string
+	dataset      string
+	backoff      backoff.Backoff
+	callbacks    Callbacks
+	setHeaders   func(r *requests.Request)
 }
 
-type Option func(c *Client) error
+type Option func(c *Client)
 
 // WithHTTPClient returns an option for setting a custom HTTP client.
 func WithHTTPClient(client *http.Client) Option {
-	return func(c *Client) error {
-		c.hc = client
-		return nil
-	}
+	return func(c *Client) { c.hc = client }
 }
 
 // WithCallbacks returns an option that enables callbacks for common events
 // such as errors.
 func WithCallbacks(cbs Callbacks) Option {
-	return func(c *Client) error {
-		c.callbacks = cbs
-		return nil
-	}
+	return func(c *Client) { c.callbacks = cbs }
 }
 
 // WithBackoff returns an option that configures network request backoff. For how
 // backoff works, see the underlying backoff package: https://github.com/jpillora/backoff.
 // By default, the client uses the backoff package's default (maximum 10 seconds wait,
 // backoff factor of 2).
 func WithBackoff(b backoff.Backoff) Option {
-	return func(c *Client) error {
-		c.backoff = b
-		return nil
-	}
+	return func(c *Client) { c.backoff = b }
 }
 
 // WithToken returns an option that sets the API token to use.
 func WithToken(t string) Option {
-	return func(c *Client) error {
-		c.token = t
-		return nil
-	}
-}
-
-// WithDataset returns an option that sets the dataset name.
-func WithDataset(id string) Option {
-	return func(c *Client) error {
-		c.dataset = id
-		return nil
-	}
+	return func(c *Client) { c.token = t }
 }
 
 // WithCDN returns an option that enables or disables the use of the Sanity API CDN.
 func WithCDN(b bool) Option {
-	return func(c *Client) error { c.useCDN = b; return nil }
+	return func(c *Client) { c.useCDN = b }
 }
 
-// WithBaseURL returns an option that changes the API URL.
-func WithBaseURL(url url.URL) Option {
-	return func(c *Client) error { c.baseURL = url; return nil }
+// WithHTTPHost returns an option that changes the API URL.
+func WithHTTPHost(scheme, host string) Option {
+	return func(c *Client) {
+		c.baseAPIURL.Scheme = scheme
+		c.baseAPIURL.Host = host
+		c.baseQueryURL.Scheme = scheme
+		c.baseQueryURL.Host = host
+	}
 }
 
-// New returns a new client. A project ID must be provided. Zero or more options can
-// be passed. For example:
+// NewClient returns a new versioned client. A project ID must be provided.
+// Zero or more options can be passed. For example:
 //
-//     client := sanity.New("foobar123",
+//     client := sanity.VersionV20210325.NewClient("projectId", sanity.DefaultDataset,
 //       sanity.WithCDN(true), sanity.WithToken("mytoken"))
 //
-func New(projectID string, opts ...Option) (*Client, error) {
+func (v Version) NewClient(projectID, dataset string, opts ...Option) (*Client, error) {
```

Comment:
I was thinking we'd keep `New()` as before, but permit you to invoke `sanity.V20210325.New()` or whatever.

Reply:
And, what should be the default API version on `New()`? Shall we keep it inline with JS client and use `v1`?

Reply:
Sounds like it's worth talking about together. My gut feeling is "newest major version" following a tagging schema (`go get github.com/sanity-io/client-go@v2021-03-25`, if that tag is possible), but release management is a pain, so I'm not sure.

Reply:
I will keep default version as `v1` (as it is today in Go and JS clients) and initiate a discussion internally about how to release and how to decide a default version in client implementations.